### PR TITLE
Introduce assorted Reactor `StepVerifier` Refaster rules

### PR DESCRIPTION
**Pull request by werli**
### Summary
This PR introduces a bunch of rules associated to `StepVerifier.Asserations` and a trivial `StepVerifier` rule.

---
### Example
I spotted the following interesting pattern:
```java
Mono.empty()
    .as(StepVerifier::create)
    .expectError(SpecificException.class)
    .verifyThenAssertThat()
    .hasOperatorErrorWithMessage("msg");
```
Which made me find the interesting `StepVerifier#Assertions` API. It's quite low-level, and IMO not as straightforward to read and write.

The example itself is most often represented in our code-base using AssertJ's richer API:
```java
Mono.empty()
    .as(StepVerifier::create)
    .verifyErrorSatisfies(t ->
        assertThat(t)
        .isInstanceOf(SpecificException.class)
        .hasMessage("msg"));
```
---
### Suggested commit message:
```
Introduce assorted Reactor `StepVerifier` Refaster rules (#1132)
```

### CODE REVIEW

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java
Comment (werli) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1970]: Asserting both class and message blocks one to use the recommended `StepVerifier.Step#verifyError(Class)` or `StepVerifier.Step#verifyErrorMessage(String)`. Choosing an alternative isn't easy since there are many possibilities. Out of all the options, AssertJ's assertions have the richest API, so I'd recommend this.

The `StepVerifierVerify` rule introduced in this PR aims to rewrite the use-cases in which only one of class or message are asserted by dropping the `StepVerifier.Assertions` type which allows other rules to rewrite further.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java
Comment (werli) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1930]: This case is not exhaustive due to the overload of `StepVerifier.LastStep#expectedError`, but IMO covers sufficiently.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java
Comment (werli) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java:1808]: The API spec mentions this is a noop. I didn't fully understand why this API may be useful for in the first place.

**Review by Stephan202 — APPROVED** [review]
Added a commit. Tnx @werli!

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java
Comment (werli) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java]: My intent was indeed to avoid breaking code changes by only matching the "dangling" case in which the created `StepVerifier.Assertions` isn't used or cases we should clearly rewrite.

`StepVerifier#verifyThenAssertThat` has some valid yet niche usages, see [internal usages](https://github.com/search?q=org%3APicnicSupermarket%20verifyThenAssertThat&type=code). Sure, we could suppress the warning here, but IMO the warning would be a false positive. WDYT?

**Review by rickie — APPROVED** [review]
nice improvements!

**Review by Venorcis — APPROVED** [review]

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java
Comment (werli) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java]: Okay - I've given this another thought. Let's proceed as-is, and violations could be suppressed through the warning.  The internal use cases are quite sparse.

**Comment by werli** [comment]
^ Rebased to resolve conflicts. IMO, we can proceed. 👍

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java]: Discussed offline: I do think @werli has a good point. Ideally we have our cake and eat it too (i.e., have separate "aggressive" and "safe" modes). I'll add a comment describing the current thinking of what that could look like.

**Review by Stephan202 — APPROVED** [review]
Added a commit, summarizing what we discussed offline. Let's merge after #1387 is merged and released.

**Comment by Stephan202** [comment]
The build fails due to what _appears_ to be a bug in JDK 23. Will require further investigation later ~today.

**Comment by Stephan202** [comment]
> The build fails due to what _appears_ to be a bug in JDK 23. Will require further investigation later ~today.

Will require attaching a debugger, I think, for which I'll need to find more time. Observations so far:
- The errors happen for `{@link StepVerifier.LastStep#verify()}` and `{@link StepVerifier.LastStep#verify(Duration)}`.
- There are _no_ errors for `{@link StepVerifier.LastStep#verifyError()}` and `{@link StepVerifier.LastStep#verifyError(Class)}`. (This is surprising: same (nested) class, also a pair of overloads.)
- As expected, the issue doesn't trigger with `-Xdoclint:-reference`, as the relevant validation is then skipped altogether.
- Based on a quick test with `22.0.1-graalce` (which I happened to still have installed locally) the issue isn't present in JDK 22.

**Review by werli — COMMENTED** [review]
The last changes  LGTM, thanks @Stephan202 💪 
Interesting bug investigation. What do we do in this scenario for `error-prone-support`? Change our code & report, or report & leave the PR open until it's fixed? The latter may take a while.

**Comment by Stephan202** [comment]
> Interesting bug investigation. What do we do in this scenario for `error-prone-support`? Change our code & report, or report & leave the PR open until it's fixed? The latter may take a while.

Discussed offline: certainly we'll want to proceed with this PR. Whatever workaround we go for, ideally we do link it to a bug report. This would likely entail the following steps:
1. Use a debugger to better understand and be able to describe the issue and, if possible, create a minimal reproducible example.
2. Create an account to file a JDK bug ticket.
3. File a bug ticket.

Depending on availability one of us will have a crack at this.

**Comment by Stephan202** [comment]
Alright, I rebased and added a commit. Turns out the bug is in all versions _preceding_ JDK 23: There are no `StepVerifier.LastStep#verify` overloads: `#verify` is defined on `StepVerifier` instead. :sweat_smile:. Will merge once built.

**Comment by werli** [comment]
:bulb::bulb::bulb: Thanks @Stephan202 💪

**Pull request merged**
